### PR TITLE
Use ScheduleLambda to schedule work on Matter stack task

### PR DESCRIPTION
**Pull request by lucicop**
Use ScheduleLambda in several places to schedule work on Matter stack for better synchronization.

### CODE REVIEW

File: examples/lighting-app/qpg/src/AppTask.cpp [hunk 1]
```diff
@@ -118,16 +118,21 @@ void OnTriggerIdentifyEffect(Identify * identify)
     case EMBER_ZCL_IDENTIFY_EFFECT_IDENTIFIER_BLINK:
     case EMBER_ZCL_IDENTIFY_EFFECT_IDENTIFIER_BREATHE:
     case EMBER_ZCL_IDENTIFY_EFFECT_IDENTIFIER_OKAY:
-        (void) chip::DeviceLayer::SystemLayer().StartTimer(chip::System::Clock::Seconds16(5), OnTriggerIdentifyEffectCompleted,
-                                                           identify);
+        SystemLayer().ScheduleLambda([identify] {
```
Comment (bzbarsky-apple) [inline examples/lighting-app/qpg/src/AppTask.cpp:121]: The documentation for ScheduleLambda says:
```
     *   Schedules a lambda object to be run as soon as possible in the CHIP context. This function is not thread-safe,
     *   it must be called with in the CHIP context
```

so how is this helping anything?

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: examples/lighting-app/qpg/src/AppTask.cpp (hunk 1 above)
Comment (bzbarsky-apple) [reply examples/lighting-app/qpg/src/AppTask.cpp:121]: Or is that documentation wrong?  In that case we should adjust the documentation....

Comment (lucicop) [reply examples/lighting-app/qpg/src/AppTask.cpp:123]: From what I saw, similar approaches exist in other example apps. In our tests, there was no issue using it - but indeed that's no guarantee. Also, there is a conflicting comment in NetworkCommissioning.h (line 247):

`         * OnFinished() must be called in a thread-safe manner with CHIP stack. (e.g. using ScheduleWork or ScheduleLambda)`

so indeed, maybe the thread safety aspect needs revisiting.

Comment (bzbarsky-apple) [reply examples/lighting-app/qpg/src/AppTask.cpp:123]: `ScheduleWork` on the `SystemLayer` is absolutely not threadsafe.

`ScheduleWork` on the `PlatformManager` is threadsafe.

Comment (lucicop) [reply examples/lighting-app/qpg/src/AppTask.cpp:123]: By looking at ScheduleLambda and PlatformEventing::ScheduleLambdaBridge, I would say they are thread safe due to posting the event to PlatformMgr queue - similar mechanism to ScheduleWork - PlatformMgr version.

Comment (bzbarsky-apple) [reply examples/lighting-app/qpg/src/AppTask.cpp:123]: Right, that's what I said above.  The question is whether any of the code before that does something weird (e.g. in its constructors, etc).

**Review by bzbarsky-apple — APPROVED** [review]
Approved, conditional on:

1. Auditing that nothing in ScheduleLambda before the ScheduleWork call touches any shared state.
2. Filing an issue to correct the documentation to match reality.

**Review by andy31415 — APPROVED** [review]

File: examples/lighting-app/qpg/src/AppTask.cpp (hunk 1 above)
Comment (lucicop) [reply examples/lighting-app/qpg/src/AppTask.cpp:123]: [ScheduleLambda_Audit.pdf](https://github.com/project-chip/connectedhomeip/files/11454697/ScheduleLambda_Audit.pdf)
Adding audit for ScheduleLambda thread safety.

**Review by bzbarsky-apple — APPROVED** [review]
This is fine as long as the CancelTimer callers are not expecting it to take effect immediately...

**Comment by lucicop** [comment]
> This is fine as long as the CancelTimer callers are not expecting it to take effect immediately...

Indeed, and I've put "this->mFunctionTimerActive = False" inside lambda with the intent that it can be used to check when the timer has actually been cancelled.

**Comment by lucicop** [comment]
Created new issue to correct documentation: https://github.com/project-chip/connectedhomeip/issues/26538

**Review by arkq — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]

**Pull request merged**